common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +177,463 @@
+                                    LOGGER.warn("load [{}] class fail. {}", line, e.getMessage());
+                                }
                             }
                         }
+                    } catch (Throwable e) {
+                        LOGGER.warn(e.getMessage());
+                    } finally {
+                        IOUtil.close(reader);
                     }
-                } catch (Throwable e) {
-                    LOGGER.warn(e.getMessage());
-                } finally {
-                    IOUtil.close(reader);
                 }
             }
         }
-    }
 
-    /**
-     * init instance
-     *
-     * @param <S>       the type parameter
-     * @param service   the service
-     * @param implClazz the impl clazz
-     * @param argTypes  the arg types
-     * @param args      the args
-     * @return s s
-     * @throws IllegalAccessException the illegal access exception
-     * @throws InstantiationException the instantiation exception
-     * @throws NoSuchMethodException the no such method exception
-     * @throws InvocationTargetException the invocation target exception
-     */
-    protected static <S> S initInstance(Class<S> service, Class implClazz, Class[] argTypes, Object[] args)
-        throws IllegalAccessException, InstantiationException, NoSuchMethodException, InvocationTargetException {
-        S s = null;
-        if (argTypes != null && args != null) {
-            // Constructor with arguments
-            Constructor<S> constructor = implClazz.getDeclaredConstructor(argTypes);
-            s = service.cast(constructor.newInstance(args));
-        } else {
-            // default Constructor
-            s = service.cast(implClazz.newInstance());
+        private ExtensionURL getExtensionURL(Class<?> clazz) {
+            String serviceName = null;
+            String typeName = null;
+            Integer priority = 0;
+            Scope scope = Scope.SINGLETON;
+            LoadLevel loadLevel = clazz.getAnnotation(LoadLevel.class);
+            typeName = clazz.getTypeName();
+            if (loadLevel != null) {
+                serviceName = loadLevel.name();
+                priority = loadLevel.order();
+                scope = loadLevel.scope();
+            }
+            ExtensionURL result = new ExtensionURL(serviceName, typeName, priority, scope);
+            //do cache
+            extensionClassUrlMap.put(clazz, result);
+            if (loadLevel != null) {
+                if (extensionNameUrlsMap.containsKey(serviceName)) {
+                    List<ExtensionURL> urls = extensionNameUrlsMap.get(serviceName);
+                    urls.add(result);
+                } else {
+                    List<ExtensionURL> urls = new ArrayList<>();
+                    urls.add(result);
+                    extensionNameUrlsMap.put(serviceName, urls);
+                }
+            }
+            if (priority >= highestLoadPriority) {
```

Comment:
Since priority defaults to 0, is the statement better to be `if (defaultExtensionURL == null || priority > highestLoadPriority)`?

Reply:
the highestLoadPriority is set to Integer.MIN_VALUE,so even if the default priority is 0, the old statement still works,but your statement will be easier to understand here.